### DISCUSSION TITLE
Cannot post new transaction due to 422 status code

### DISCUSSION THREAD

**Discussion by bertbesser** [firefly-iii/firefly-iii#4584] [Firefly III] [answered]
I am running Firefly III version 5.4.6

**Description**
Hi,

I'm trying to

```
curl -L --header "Authorization: Bearer `cat ffiii.pat`" --header 'Accept: application/json' --header "Content-Type: application/json" --request POST --data "`cat import.json`" 'http://my-host/api/v1/transactions'
```

where `ffiii.pat` and `import.json` contain, respectively, my PAT and

```
{
  "transactions": [
    {
      "type": "withdrawal",
      "date": "2016-12-28",
      "amount": "6.01",
      "description": "no comment",
      "category_name": "groceries",
      "source_id": 1,
      "destination_id": 7
    }
  ]
}
```

I get this answer;

```
{
  "message": "The given data was invalid.",
  "errors": {
    "transactions.0.description": [
      "Need at least one transaction.",
      "The description field is required."
    ],
    "transactions.0.type": [
      "Invalid transaction type."
    ]
  }
}
```

and the server logs show a `422` status code.

What am I missing?

Thanks a bunch!

**Extra info**
 
the about API endpoint shows version 1.4.0

**Comment by JC5** [comment]
What's the output of `echo [your command]`?

When I run your code it works. 

<img width="690" alt="Screenshot 2021-04-02 at 19 20 35" src="https://user-images.githubusercontent.com/5889984/113438270-c3f81800-93d7-11eb-8077-b33be235029d.png">

**Comment by bertbesser** [comment]
thx for responding so quickly

🤔 
```
echo curl -L --header "Authorization: Bearer `cat ffiii.pat`" --header 'Accept: application/json' --header "Content-Type: application/json" --request POST --data "`cat import.json`" 'http://myhost/api/v1/transactions'
curl -L --header Authorization: Bearer ey[shortened] --header Accept: application/json --header Content-Type: application/json --request POST --data {
  "transactions": [
    {
      "type": "withdrawal",
      "date": "2016-12-28T12:00:00+01:00",
      "amount": "6.01",
      "description": "no comment",
      "category_name": "groceries",
      "source_id": 1,
      "destination_id": 7
    }
  ]
} http://myhost/api/v1/transactions
```

// slight changes since i changed the date format in the meantime

are there some hidden API configurations (such as denying write requests or similar) that i might miss?

btw: i'm on docker

**Comment by JC5** [comment]
> are there some hidden API configurations (such as denying write requests or similar) that i might miss?

Nope. And they wouldn't result in a 422 in any case. I'm working on something else atm, I'll do some digging later.

**Comment by bertbesser** [comment]
No rush. I'll fiddle with adding transactions in the coming days. Will report back. Happy easter!

**Comment by JC5** [comment]
I've been playing around with Postman. The JSON content at least is correct. Since the expand seems to work fine I'm not sure what the issue could be. All I can suggest is backup, upgrade, and try again.

**Comment by M-Paladin** [comment]
I tried the same thing using curl on my server and using rester on my client with firefox and I have the same problem
I tried many different syntax based on the doc but for the moment, nothing seems to work

Edit : Oh my bad, I had an extra comma at the end of the transaction so yes it's also working for me from server with curl as well as client firefox with rester
I used the same syntax as you (adapted with my case) but the difference is I'm in 5.5.2

**Comment by bertbesser** [comment, marked as answer]
I tried again yesterday and it worked as expected. So it was some mistake on my side. I'm pretty sure that it was due to wrong quotation, i.e. using single quotes like in
```
... --data '`cat import.json`' ...
```
does _not_ execute the `cat` command, whereas using double quotes does:
```
... --data "`cat import.json`" ...
```
I.e., only in the latter example fireflyIII receives a JSON request body.

This source of errors can be prevented by reading data directly from a file like so:
```
... --data @import.json ...
```

Sorry for all the confusion and thanks again for investigating.

**Comment by M-Paladin** [comment]
You can also use this --data "$(cat import.json)" which is best practice according to https://www.shellcheck.net/

**Comment by Findus23** [comment]
In case it helps anyone:
I was getting the same issue when I missed the `"Content-Type": "application/json"` header in the request.
